### ISSUE TITLE
Unauthorized: HTTP status code 401; transport: received the unexpected content-type "text/html; charset=UTF-8"

### ISSUE THREAD

**Issue by DazWilkin** [GoogleCloudPlatform/esp-v2#542]
I'm trying to configure a gRPC-based service behind Cloud Endpoints with Firebase Auth.

Two Google Projects:

1. Firebase Authentication is running in `FIREBASE_PROJECT`
1. Cloud Endpoints and the Cloud Run gRPC service `PROJECT`

But, using gRPCurl and passing the JWT from Firebase Authentication in a `Authorization` header, I get 401s:

```bash
grpcurl \
-v  \
-H "Authorization: Bearer ${TOKEN}" \
-d "{ ... }" -\
proto some.proto \
${ESP_HOST}:${PORT} \
v1alpha1.SomeService/SomeMethod

Resolved method descriptor:
rpc SomeMethod ( .v1alpha1.SomeMethodGetRequest ) returns ( .v1alpha1.SomeMethodGetResponse );

Request metadata to send:
(empty)

Response headers received:
(empty)

Response trailers received:
(empty)
Sent 1 request and received 0 responses
ERROR:
  Code: Unauthenticated
  Message: Unauthorized: HTTP status code 401; transport: received the unexpected content-type "text/html; charset=UTF-8"
```

And, I'm unable to find much useful in the logs and I'm unable to find `X-Endpoint-API-UserInfo` or `Authorization` headers:

```JSON
{
  "textPayload": "The request was not authorized to invoke this service. Read more at https://cloud.google.com/run/docs/securing/authenticating",
  "insertId": "60c9055e0006b28a8688e19c",
  "httpRequest": {
    "requestMethod": "POST",
    "requestUrl": "https://proxy-[[hash]]-uw.a.run.app/v1alpha1.SomeService/SomeMethod",
    "requestSize": "1515",
    "status": 401,
    "userAgent": "grpc-go/1.31.0",
    "remoteIp": "66.235.10.167",
    "serverIp": "216.239.36.53",
    "latency": "0s",
    "protocol": "HTTP/1.1"
  },
  "resource": {
    "type": "cloud_run_revision",
    "labels": {
      "revision_name": "proxy-00008-pif",
      "project_id": "[[project]]",
      "configuration_name": "proxy",
      "location": "us-west1",
      "service_name": "proxy"
    }
  },
  "timestamp": "2021-06-15T19:54:06.438922Z",
  "severity": "WARNING",
  "logName": "projects/[[project]]/logs/run.googleapis.com%2Frequests",
  "receiveTimestamp": "2021-06-15T19:54:06.590759399Z"
}
```
I don't always get a corresponding error in the Endpoints Logs but here's an earlier one:

```JSON
{
  "textPayload": "The request was not authorized to invoke this service. Read more at https://cloud.google.com/run/docs/securing/authenticating",
  "insertId": "60c9055e0006b28a8688e19c",
  "httpRequest": {
    "requestMethod": "POST",
    "requestUrl": "https://proxy-[[hash]]-uw.a.run.app/v1alpha1.SomeService/SomeMethod",
    "requestSize": "1515",
    "status": 401,
    "userAgent": "grpc-go/1.31.0",
    "remoteIp": "66.235.10.167",
    "serverIp": "216.239.36.53",
    "latency": "0s",
    "protocol": "HTTP/1.1"
  },
  "resource": {
    "type": "cloud_run_revision",
    "labels": {
      "revision_name": "proxy-00008-pif",
      "project_id": "[[project]]",
      "configuration_name": "proxy",
      "location": "us-west1",
      "service_name": "proxy"
    }
  },
  "timestamp": "2021-06-15T19:54:06.438922Z",
  "severity": "WARNING",
  "logName": "projects/[[project]]/logs/run.googleapis.com%2Frequests",
  "receiveTimestamp": "2021-06-15T19:54:06.590759399Z"
}
```

I've used [`jwt.io`](https://jwt.io) to decode the JWT and the payload appears correct, e.g.:

```JSON
{
  "name": "Daz Wilkin",
  "iss": "https://securetoken.google.com/[[FIREBASE_PROJECT]]",
  "aud": "[[FIREBASE_PROJECT]]",
  "auth_time": 1623786138,
  "user_id": "...",
  "sub": "...",
  "iat": 1623786138,
  "exp": 1623789738,
  "email": "[[EMAIL]]",
  "email_verified": true,
  "firebase": {
    "identities": {
      "microsoft.com": [
        "29d79103b80d51a0"
      ],
      "email": [
        "[[EMAIL]]"
      ]
    },
    "sign_in_provider": "microsoft.com"
  }
}
```
I've tried adding the Cloud Endpoints|Run project's (`PROJECT`) default compute service account (as the service that's running Endpoints and the Run services) to the Firebase project (`FIREBASE_PROJECT`) thinking perhaps it needed to auth but this makes no difference.

I noticed that the JWT's `aud` is not of the form `https://SERVICE_NAME` (per Google's documents) but is just `SERVICE_NAME` so I tried including the `audience` property of `authentication` but it appears to make no difference.

The configuration works without authentication; i.e. when both Cloud Run services are deployed with `--allow-unauthenticated` and I don't include an `Authorization` header:

```bash
grpcurl \
-v  \
-d "{ ... }" -\
proto some.proto \
${ESP_HOST}:${PORT} \
v1alpha1.SomeService/SomeMethod

Resolved method descriptor:
rpc SomeMethod ( .v1alpha1.SomeMethodGetRequest ) returns ( .v1alpha1.SomeMethodGetResponse );

Request metadata to send:
(empty)

Response headers received:
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
content-type: application/grpc
date: Tue, 15 Jun 2021 19:34:13 GMT
server: Google Frontend
x-envoy-decorator-operation: ingress SomeMethod

Response contents:
{
  ...
}

Response trailers received:
(empty)
Sent 1 request and received 1 response
```
If I add the `authentication` section:
```YAML
type: google.api.Service
config_version: 3
name: proxy-[[hash]]-uw.a.run.app
title: Some Service
apis:
  - name: v1alpha1.SomeService
usage:
  rules:
    - selector: "*"
      allow_unregistered_calls: true
backend:
  rules:
    - selector: "*"
      address: grpcs://service-[[hash]]-uw.a.run.app:443
authentication:
  providers:
  - id: firebase
    issuer: https://securetoken.google.com/[[FIREBASE_PROJECT]]
    jwks_uri: https://www.googleapis.com/service_accounts/v1/metadata/x509/securetoken@system.gserviceaccount.com
    audiences: [[FIREBASE_PROJECT]]  # Tried with|without this
  rules:
  - selector: "*"
    requirements:
      - provider_id: firebase
```

**Comment by nareddyt** [comment]
Can you summarize your deployment architecture? Is it like the diagram on this page? https://cloud.google.com/endpoints/docs/openapi/get-started-cloud-run

If so, what Cloud Run authentication did you configure on the ESPv2 instance? You should deploy ESPv2 Cloud Run with `--allow-unauthenticated`. Anyone can reach the ESPv2 instance, but ESPv2 will verify the Firebase Auth token and only let authenticated requests to the backend.

The backend Cloud Run instance should be configured to only allow ESPv2 call it. This is described in step 4 of [this section](https://cloud.google.com/endpoints/docs/openapi/get-started-cloud-run#deploy_esp).

**Comment by qiwzhang** [comment]
It seems that error is from Cloud Run, not from ESPv2
"The request was not authorized to invoke this service. Read more at https://cloud.google.com/run/docs/securing/authentication"

When the error happens,  did you "gcloud run deploy" ESPv2 service with `--allow-unauthenticated` flag?  If no, that may explain it

You should deploy ESPv2 service with with `--allow-unauthenticated` flag in order to let ESPv2 verify "Authorization" header.

**Comment by DazWilkin** [comment]
Thank you both!

User error.... I was running the proxy with `--no-allow-unauthenticated`.

After redeploying the proxy with `--allow-unauthenticated`, the next, authenticated request suceeded:

```bash
grpcurl \
-v  \
-d "{ ... }" -\
proto some.proto \
${ESP_HOST}:${PORT} \
v1alpha1.SomeService/SomeMethod

Resolved method descriptor:
rpc SomeMethod ( .v1alpha1.SomeMethodGetRequest ) returns ( .v1alpha1.SomeMethodGetResponse );

Request metadata to send:
authorization: Bearer eyJhbGci...dIHg

Response headers received:
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
content-type: application/grpc
date: Wed, 16 Jun 2021 16:31:34 GMT
server: Google Frontend
x-envoy-decorator-operation: ingress SomeMethod

Response contents:
{
  "foo": {
    "user": "Daz Wilkin",
    "baz": "..."
  }
}

Response trailers received:
(empty)
Sent 1 request and received 1 response
```

Thank you very much!

**Comment by DazWilkin** [comment]
One additional note, I am having to include the `authentication.provider.audiences` (string) in the service config because Firebase JWT's `aud` property is the Firebase Project ID:

|claim|value|
|-----|-----|
|`aud`|`${FIREBASE_PROJECT}`|
|`exp`|`1623864483`|
|`iat`|`1623860883`|
|`iss`|`https://securetoken.google.com/d${FIREBASE_PROJECT}`|
|`sub`|`[[REDACTED]]`|
|`user_id`|`[[REDACTED]]`|

Google's [documentation](https://cloud.google.com/endpoints/docs/grpc/authenticating-users#firebase) has:

```YAML
# Optional. Replace YOUR-CLIENT-ID with your client ID
audiences: "YOUR-CLIENT-ID"
```
And: "The audiences field isn't required. ESP accepts all JWTs with the back-end service name in the form of https://SERVICE_NAME in the aud claim. To permit additional client IDs to access the back-end service, you can specify the allowed client IDs in the audiences field by using comma-separated values. ESP then accepts the JWTs with the whitelisted client IDs in the aud claim."

But, my experience is that, unless I have the following, I'll get 403s from the proxy:

```YAML
type: google.api.Service
config_version: 3
...
authentication:
  providers:
    - id: firebase
      issuer: https://securetoken.google.com/${FIREBASE_PROJECT}
      jwks_uri: https://www.googleapis.com/service_accounts/v1/metadata/x509/securetoken@system.gserviceaccount.com
      audiences: ${FIREBASE_PROJECT}
...
```

**Comment by qiwzhang** [comment]
That is correct, works as intended.  "aud" in the JWT token has to match either 1) https://SERVICE_NAME  if `authentication.provider.audiences` is empty.  or 2) whatever specified in `authentication.provider.audiences`.

Since your "aud" in JWT token is FIREBASE_PROJECT,   you have to use 2).

**Comment by DazWilkin** [comment]
I understand but, the documentation is specifically for Firebase and it says that `audiences` is optional and it is not because the Firebase JWT does not use `https://SERVICE_NAME`.

I'm proposing that the documentation be updated to read that the `audience` value is mandatory (not optional) and needs to reflect the Firebase Project ID.

The documentation also uses the term OAuth term "Client ID" but Firebase does not use this terminology and refers to this as the Project ID.

**Comment by qiwzhang** [comment]
I see.  will update the doc.  Thanks.
